rs/rust_canisters/random_traffic_test/src/main.rs

```diff
@@ -0,0 +1,236 @@
+    let payload_bytes = gen_range(|config| config.call_bytes_min..=config.call_bytes_max);
+    let call_id = next_call_id();
+    let insert_new_call_record = |reply: Option<Reply>| {
+        RECORDS.with_borrow_mut(|records| {
+            records.insert(
+                call_id,
+                Record {
+                    receiver,
+                    sent_bytes: payload_bytes,
+                    reply,
+                },
+            );
+        });
+    };
+    let set_reply_in_call_record = move |reply: Reply| {
+        RECORDS.with_borrow_mut(|records| {
+            records.get_mut(&call_id).unwrap().reply = Some(reply);
+        });
+    };
+
+    let error_code = api::call_with_callbacks(
+        receiver,
+        "handle_call",
+        &vec![0_u8; payload_bytes as usize][..],
+        move || {
+            set_reply_in_call_record(Reply::Bytes(api::arg_data().len() as u32));
+            on_response();
+        },
+        move || {
+            set_reply_in_call_record(Reply::AsynchronousRejection(
+                api::reject_code(),
+                api::reject_message(),
+            ));
+            on_response();
+        },
+    );
+    if error_code == 0 {
+        insert_new_call_record(None);
+        Ok(())
+    } else {
+        insert_new_call_record(Some(Reply::SynchronousRejection(error_code)));
+        Err(())
+    }
+}
+
+/// Samples a weighted binomial distribution to decide whether to make a reply (true) or a
+/// downstream call (false). Defaults to `true` for bad weights (e.g. both 0).
+fn should_reply() -> bool {
+    RNG.with_borrow_mut(|rng| {
+        WeightedIndex::new([REPLY_WEIGHT.get(), CALL_WEIGHT.get()])
+            .map(|dist| dist.sample(rng) == 0)
+            .unwrap_or(true)
+    })
+}
+
+/// Handles incoming calls; this method is called from the heartbeat method.
+///
+/// Replies if:
+/// - sampling the weighted binomial distribution tells us to do so.
+/// - if it tells us not to do so but the attempted downstream call fails synchronously.
+///
+/// Note that a reply to a successful downstream call is made from the callbacks provided for this
+/// call; this amounts to awaiting the outcome of the call asynchronously first.
+#[export_name = "canister_update handle_call"]
+fn handle_call() {
+    // Passing handle_call() as the `on_response` for `try_call()` leads to random call trees where
+    // it is recursively decided to make another call or reply on each node upon awaiting the
+    // outcome. The shape of the call tree is statistically determined by the values of `REPLY_WEIGHT`
+    // and `CALL_WEIGHT`.
+    //
+    // Note: Awaiting each call before possibly making another one also ensures there is only ever
+    // one response despite potentially multiple downstream calls, something that doesn't matter for
+    // the heartbeat where calls are not awaited because the heartbeat can't give a response anyway.
+    if should_reply() || try_call(handle_call).is_err() {
```

Comment:
It's quite hard to understand what's going on here: on the one hand, not only do we decide whether or not to make a downstream call before replying (resulting in a call chain of arbitrary length), but we also go through the same process when we get the response (by abusing `handle_call()` as a response handler). Which still only ever gives us a call chain (we never end up with concurrent downstream calls), only with a hard to pin down distribution (essentially we flip the coin twice on every call, with a time skew involved, so that it all looks like an in-depth traversal of a call tree).

Why not simply decide whether to make a downstream call or not; if no downstream call, directly produce a reply; if we do decide to first make a downstream call, produce the reply in the callback(s)?

Reply:
- The length is not arbitrary, it is random but statistically determined by `REPLY_WEIGHT` and `CALL_WEIGHT` (i.e. `should_reply_now()` returns `true`) and also capped by the available message memory (i.e. try_call() return `Err()`); one or the other is bound to happen at some point, so each call tree must eventually terminate.
- This is not abuse, but perhaps it should be called something other than `handle_call()`.
- The whole point is to make this random. At each node in the tree we can either reply or make a downstream call. Once this call concludes traversing whatever subtree it may have spawned downstream and comes back to this node, we again decide whether to make another downstream call or else reply. This way we get both a random depth and a random number of branches on each node.

> if we do decide to first make a downstream call, produce the reply in the callback(s)?
This is what it does, except that I need a way to turn the reply on and off depending on whether the call was made by the heartbeat or something further down. Replying on the heartbeat traps the canister, hence the no-op passed to `try_call()`.

Reply:
> This way we get both a random depth and a random number of branches on each node.

What does the latter give us that the former doesn't? I.e. what is the difference between repeatedly deciding whether to make downstream calls; and adjusting the probability of making one single downstream call? (The latter of which would, I would argue, give you simpler code structure than using (not really) recursion as a way of achieving iterative behavior.)

> > if we do decide to first make a downstream call, produce the reply in the callback(s)?
>
> This is what it does

Not exactly. In the callback, instead of directly producing a reply, you call again the request handler. Which may or may not immediately produce a reply. And I'm arguing that there's no need for this recursion. Sure, you need to pass a boolean flag as to whether to produce a reply (when handling an actual call) or not (when in a heartbeat). But that's a lot easier to follow than untangling `try_call(handle_call)`.

Reply:
It gives me random call trees with a random number of branches on each node. Without it I get nodes with only one branch. I can also use a `bool` which I considered doing but then it's just `if this_bool { handle_call() }` how is that better?

Reply:
> It gives me random call trees with a random number of branches on each node.

Conceptually, yes. But if you take a snapshot of all these call trees at any given time, every single one of them is a list of calls. So from the point of view of what this may do to memory usage; or reject signals; or anything else that we plan to test using this canister; it makes zero difference. I'm not even sure that multiple concurrent calls would make any significant difference: the heartbeat already makes multiple concurrent calls.

Reply:
What I was aiming to do here is make a canister that produces random traffic without assuming anything about the underlying implementation. In my view that should include all sorts of call patterns. If this produced a large amount of difficult code I would agree but it really doesn't. The slight awkwardness of this added recursion is entirely because I am using the low level api. All it means is 'handle the call again just like before when you are back at this point'. It is otherwise the same as using `await` in a loop where we randomly decide the stop and reply.